Add OVAL check for rule ensure_shadow_group_empty

dodys opened the pull request:
#### Description:

- Add OVAL check for rule ensure_shadow_group_empty
- Also improve remediation for same rule

#### Rationale:

- This OVAL was missing and is needed for CIS on both Ubuntu 20.04 and 22.04 as well as SUSE.

teacup-on-rockingchair commented:
> @teacup-on-rockingchair this oval is probably of SUSE's interest as well.

Indeed @dodys , thanks a lot for the heads up. I did initially tested it and it seems that I am having some issue with the case when there is a user added to the shadow group, say
```
# cat /etc/group|grep shadow
shadow:x:15:vnc
```
In that case the `ste_shadow_group_members` state does not work for me to match it as an empty string ( my guess)

Only when I modified it to :
```
  <ind:textfilecontent54_state id="ste_shadow_group_members" version="1">
      <ind:subexpression operation="pattern match">^\s*$</ind:subexpression>
  </ind:textfilecontent54_state>
```
, it caught the vnc user as a problem.
Hopefully I understood everything correctly and the above is helpful ;)
In any case :bow: for the new rule !

dodys commented:
> > @teacup-on-rockingchair this oval is probably of SUSE's interest as well.
> 
> Indeed @dodys , thanks a lot for the heads up. I did initially tested it and it seems that I am having some issue with the case when there is a user added to the shadow group, say
> 
> ```
> # cat /etc/group|grep shadow
> shadow:x:15:vnc
> ```
> 
> In that case the `ste_shadow_group_members` state does not work for me to match it as an empty string ( my guess)
> 
> Only when I modified it to :
> 
> ```
>   <ind:textfilecontent54_state id="ste_shadow_group_members" version="1">
>       <ind:subexpression operation="pattern match">^\s*$</ind:subexpression>
>   </ind:textfilecontent54_state>
> ```
> 
> , it caught the vnc user as a problem. Hopefully I understood everything correctly and the above is helpful ;) In any case bow for the new rule !

thanks @teacup-on-rockingchair 
I've fixed it and I also changed the remediation, let me know your thoughts on it

teacup-on-rockingchair commented:
> > > @teacup-on-rockingchair this oval is probably of SUSE's interest as well.
> > 
> > 
> > Indeed @dodys , thanks a lot for the heads up. I did initially tested it and it seems that I am having some issue with the case when there is a user added to the shadow group, say
> > ```
> > # cat /etc/group|grep shadow
> > shadow:x:15:vnc
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > In that case the `ste_shadow_group_members` state does not work for me to match it as an empty string ( my guess)
> > Only when I modified it to :
> > ```
> >   <ind:textfilecontent54_state id="ste_shadow_group_members" version="1">
> >       <ind:subexpression operation="pattern match">^\s*$</ind:subexpression>
> >   </ind:textfilecontent54_state>
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > , it caught the vnc user as a problem. Hopefully I understood everything correctly and the above is helpful ;) In any case bow for the new rule !
> 
> thanks @teacup-on-rockingchair I've fixed it and I also changed the remediation, let me know your thoughts on it

Looks great, thanks :bow:

marcusburghardt reviewed (commented):
It would be great to include test scenarios.

marcusburghardt commented:
Automatus CS8, CS9 and Fedora tests are failing because the rule is restricted to ubuntu and sle.

marcusburghardt commented on linux_os/guide/system/accounts/accounts-restrictions/account_expiration/ensure_shadow_group_empty/oval/shared.xml:
```diff
@@ -0,0 +1,48 @@
+    </criteria>
+  </definition>
+
+  <ind:textfilecontent54_test check="all" check_existence="only_one_exists"
```
If the system has no `shadow` group. For example, if the `root` group is used instead, this rule will fail.
The intention of the rule is to ensure the `shadow` group is empty. If the `shadow` group does not exist, the rule should pass. Maybe using `check_existence="any_exist"` should work.

dodys replied:
it this a valid case though?
If you want shadowed passwords I believe it might be mandatory to have the shadow group.
might need to dig some man pages about this

marcusburghardt replied:
I couldn't check the sle and ubuntu distros, but in RHEL, for example, there is no `shadow` group. The `/etc/shadow` is group-owned by `root` in this case. My concern is more about the robustness of the rule if anything change in the OS, for example.

marcusburghardt commented on linux_os/guide/system/accounts/accounts-restrictions/account_expiration/ensure_shadow_group_empty/tests/shadow_not_empty.fail.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/bash
+# platform = multi_platform_sle,multi_platform_ubuntu
+
+sed -i '/^shadow:[^:]*:[^:]*:/ s/$/vnc/' /etc/group
```
It would be interesting to ensure the `shadow` group exists for this test.

marcusburghardt commented on linux_os/guide/system/accounts/accounts-restrictions/account_expiration/ensure_shadow_group_empty/tests/shadow_user_member.fail.sh:
```diff
@@ -0,0 +1,6 @@
+#!/bin/bash
+# platform = multi_platform_sle,multi_platform_ubuntu
+# remediation = none
```
I believe it is fine to remediate this test.

dodys replied:
it will fail, we don't touch the `/etc/passwd` file in the remediation

marcusburghardt replied:
Got it. Thanks.

marcusburghardt reviewed (changes requested):
Thanks to the test scenarios, it was noticed some points about the group existence.

marcusburghardt commented on linux_os/guide/system/accounts/accounts-restrictions/account_expiration/ensure_shadow_group_empty/bash/shared.sh:
```diff
@@ -1,4 +1,4 @@
-grep '^shadow:[^:]*:[^:]*:[^:]+' /etc/group
-awk -F: '($4 == "<shadow-gid>") { print }' /etc/passwd
+sed -ri 's/(^shadow:[^:]*:[^:]*:)([^:]+$)/\1/' /etc/group
+awk -F: -v GID="$(awk -F: '($1=="shadow") {print $3}' /etc/group)" '($4==GID) {print $1}' /etc/passwd
```
I believe this `awk` command is very useful for checking, but it seems unnecessary in the remediation since it is only showing if any user has the `shadow` as primary group. Would it make sense to change the primary group of these users, if any is found? Otherwise we should include a warning in the `rule.yml` informing this situation.

marcusburghardt replied:
Also a test for this scenario would be interesting.

dodys replied:
> I believe this `awk` command is very useful for checking, but it seems unnecessary in the remediation since it is only showing if any user has the `shadow` as primary group. Would it make sense to change the primary group of these users, if any is found? Otherwise we should include a warning in the `rule.yml` informing this situation.

I think changing any users primary group is too disruptive.
I can remove the line.

dodys replied:
> Also a test for this scenario would be interesting.

not sure I understood what scenario you are referring to.
We already have a test for it: `shadow_user_member.fail.sh`  for users with primary group shadow

dodys replied:
I've removed the line you mentioned

marcusburghardt commented on linux_os/guide/system/accounts/accounts-restrictions/account_expiration/ensure_shadow_group_empty/oval/shared.xml:
```diff
@@ -0,0 +1,48 @@
+    {{{ oval_metadata("Ensure shadow group is empty") }}}
+    <criteria>
+      <criterion test_ref="tst_shadow_group_empty" comment="shadow group is empty" />
+      <criterion test_ref="tst_no_user_assigned_shadow_group" comment="no user is assigned to the shadow group" />
```
Besides the line length, could the comment be more specific by informing the test is actually checking the primary group of users. Maybe `no user has the shadow as primary group`.

dodys replied:
thanks for the suggestion, added it now

marcusburghardt reviewed (changes requested):
The tests are passing fine now. I only noticed and commented about some extra cases to consider.

marcusburghardt reviewed (approved):
LGTM. Thanks

marcusburghardt commented:
Automatus CS8 and CS9 are failing because this rule is restricted to `ubuntu` and `sle`.
